### ISSUE TITLE
bbc segfaults: assignment to allocatable not implemented

### ISSUE THREAD

**Issue by oroppas** [flang-compiler/f18-llvm-project#826]
`bbc` crashes on the following code:

```fortran
      PROGRAM CCHKEE

      INTEGER            NMAX
      PARAMETER          ( NMAX = 132 )
      INTEGER            NEED
      PARAMETER          ( NEED = 14 )

      INTEGER AllocateStatus
      COMPLEX, DIMENSION(:,:), ALLOCATABLE :: A

      ALLOCATE ( A(NMAX*NMAX,NEED), STAT = AllocateStatus )

      A = 0.0

      END
```

```
error: loc("./allocatable.f90":13:7): /home/ryuta/packages/llvm/flang/f18-llvm-project/flang/lib/Lower/Bridge.cpp:1686: not yet implemented assignment to allocatable not implemented
LLVM ERROR: aborting
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /home/ryuta/packages/llvm/flang/build/bin/bbc allocatable.f90
 #0 0x00007f11bba76c73 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/ryuta/packages/llvm/flang/build/lib/libLLVMSupport.so.13git+0x229c73)
 #1 0x00007f11bba7457e llvm::sys::RunSignalHandlers() (/home/ryuta/packages/llvm/flang/build/lib/libLLVMSupport.so.13git+0x22757e)
 #2 0x00007f11bba772ef SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f11bfbb4630 __restore_rt sigaction.c:0:0
 #4 0x00007f11bb080387 raise (/lib64/libc.so.6+0x36387)
 #5 0x00007f11bb081a78 abort (/lib64/libc.so.6+0x37a78)
 #6 0x00007f11bb996833 llvm::report_fatal_error(llvm::Twine const&, bool) (/home/ryuta/packages/llvm/flang/build/lib/libLLVMSupport.so.13git+0x149833)
 #7 0x00007f11bb996698 (/home/ryuta/packages/llvm/flang/build/lib/libLLVMSupport.so.13git+0x149698)
 #8 0x00007f11bf88c752 (/home/ryuta/packages/llvm/flang/build/lib/libFortranLower.so.13git+0x1da752)
 #9 0x00007f11bf8ce4f0 decltype(auto) std::__1::__variant_detail::__visitation::__base::__dispatcher<0ul>::__dispatch<std::__1::__variant_detail::__visitation::__variant::__value_visitor<Fortran::common::visitors<(anonymous namespace)::FirConverter::genAssignment(Fortran::evaluate::Assignment const&)::'lambda'(Fortran::evaluate::Assignment::Intrinsic const&), (anonymous namespace)::FirConverter::genAssignment(Fortran::evaluate::Assignment const&)::'lambda'(Fortran::evaluate::ProcedureRef const&), (anonymous namespace)::FirConverter::genAssignment(Fortran::evaluate::Assignment const&)::'lambda'(std::__1::vector<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, std::__1::allocator<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > > > const&), (anonymous namespace)::FirConverter::genAssignment(Fortran::evaluate::Assignment const&)::'lambda'(std::__1::vector<std::__1::pair<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > >, std::__1::allocator<std::__1::pair<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > > > > const&)> >&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Fortran::evaluate::Assignment::Intrinsic, Fortran::evaluate::ProcedureRef, std::__1::vector<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, std::__1::allocator<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > > >, std::__1::vector<std::__1::pair<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > >, std::__1::allocator<std::__1::pair<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > > > > > const&>(std::__1::__variant_detail::__visitation::__variant::__value_visitor<Fortran::common::visitors<(anonymous namespace)::FirConverter::genAssignment(Fortran::evaluate::Assignment const&)::'lambda'(Fortran::evaluate::Assignment::Intrinsic const&), (anonymous namespace)::FirConverter::genAssignment(Fortran::evaluate::Assignment const&)::'lambda'(Fortran::evaluate::ProcedureRef const&), (anonymous namespace)::FirConverter::genAssignment(Fortran::evaluate::Assignment const&)::'lambda'(std::__1::vector<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, std::__1::allocator<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > > > const&), (anonymous namespace)::FirConverter::genAssignment(Fortran::evaluate::Assignment const&)::'lambda'(std::__1::vector<std::__1::pair<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > >, std::__1::allocator<std::__1::pair<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > > > > const&)> >&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Fortran::evaluate::Assignment::Intrinsic, Fortran::evaluate::ProcedureRef, std::__1::vector<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, std::__1::allocator<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > > >, std::__1::vector<std::__1::pair<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > >, std::__1::allocator<std::__1::pair<Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> >, Fortran::evaluate::Expr<Fortran::evaluate::Type<(Fortran::common::TypeCategory)0, 8> > > > > > const&) Bridge.cpp:0:0
#10 0x00007f11bf8cd254 (anonymous namespace)::FirConverter::genAssignment(Fortran::evaluate::Assignment const&) Bridge.cpp:0:0
#11 0x00007f11bf8c2c43 (anonymous namespace)::FirConverter::genFIR(Fortran::lower::pft::Evaluation&, bool) Bridge.cpp:0:0
#12 0x00007f11bf8c0b21 (anonymous namespace)::FirConverter::lowerFunc(Fortran::lower::pft::FunctionLikeUnit&) Bridge.cpp:0:0
#13 0x00007f11bf8adb38 Fortran::lower::LoweringBridge::lower(Fortran::parser::Program const&, Fortran::semantics::SemanticsContext const&) (/home/ryuta/packages/llvm/flang/build/lib/libFortranLower.so.13git+0x1fbb38)
#14 0x000000000021482a main (/home/ryuta/packages/llvm/flang/build/bin/bbc+0x21482a)
#15 0x00007f11bb06c555 __libc_start_main (/lib64/libc.so.6+0x22555)
#16 0x0000000000210699 _start (/home/ryuta/packages/llvm/flang/build/bin/bbc+0x210699)
```

**Comment by jeanPerier** [comment]
This TODO is being worked upon.

**Comment by jeanPerier** [comment]
Implemented in fir-dev by #836 with conditional reallocation in assignments to whole allocatables.
Program given in the issue description now compiles and runs.
